File: innomcp-node/antigravity_watcher.py
import subprocess
import time
import datetime
import sys
import threading
import os
import signal
import io

# Force UTF-8 for Windows console
sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')
sys.stderr = io.TextIOWrapper(sys.stderr.buffer, encoding='utf-8')

# --- CONFIGURATION ---
TEST_COMMAND = ["npx", "jest", "tests/weather_regression_phase65_final.test.ts"]  # Target specific regression suite
MAX_DURATION_SECONDS = 6 * 3600  # 6 Hours
IDLE_TIMEOUT_SECONDS = 10 * 60   # 10 Minutes
RETRY_DELAY_SECONDS = 10         # Delay before retrying after failure
MAX_RETRIES = 1000               # Effective infinite retries for 6h duration

LOG_FILE = "antigravity_session.log"
REPORT_FILE = "summary_report.md"

# --- GLOBAL STATE ---
start_time = time.time()
running = True
last_output_time = time.time()
process = None
passed_specs = []
failed_specs = []
stuck_events = 0
restarts = 0

def log(message):
    """Writes to console and log file with timestamp."""
    timestamp = datetime.datetime.now().strftime("[%Y-%m-%d %H:%M:%S]")
    line = f"{timestamp} {message}"
    print(line)
    with open(LOG_FILE, "a", encoding="utf-8") as f:
        f.write(line + "\n")

# --- LOG CLEANER THREAD ---
class LogCleaner(threading.Thread):
    def __init__(self, filepath, max_lines=5000, check_interval=30):
        super().__init__()
        self.filepath = filepath
        self.max_lines = max_lines
        self.check_interval = check_interval
        self.daemon = True
        self.stop_event = threading.Event()

    def run(self):
        log(f"🧹 LogCleaner started. Monitoring {self.filepath} (Max: {self.max_lines} lines)")
        while not self.stop_event.is_set():
            try:
                self.clean()
            except Exception as e:
                log(f"⚠️ LogCleaner error: {e}")
            
            self.stop_event.wait(self.check_interval)

    def clean(self):
        if not os.path.exists(self.filepath):
            return

        try:
            # Check size first to avoid reading small files
            if os.path.getsize(self.filepath) < 1024 * 1024: # < 1MB
                return

            with open(self.filepath, "r", encoding="utf-8", errors='ignore') as f:
                lines = f.readlines()

            if len(lines) > self.max_lines:
                # Keep last 50%
                keep_count = int(self.max_lines * 0.5)
                new_lines = lines[-keep_count:]
                
                with open(self.filepath, "w", encoding="utf-8") as f:
                    f.writelines(new_lines)
                
                log(f"🧹 Log cleaned: Reduced from {len(lines)} to {len(new_lines)} lines.")
        except Exception as e:
            # Don't log here to avoid infinite loop if logging fails
            print(f"Log clean error: {e}")

    def stop(self):
        self.stop_event.set()

log_cleaner = None

def generate_report():
    """Generates the final markdown report."""
    duration = time.time() - start_time
    hours = int(duration // 3600)
    minutes = int((duration % 3600) // 60)
    
    with open(REPORT_FILE, "w", encoding="utf-8") as f:
        f.write(f"# 🛡️ Antigravity Watcher Report\n\n")
        f.write(f"**Date:** {datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n")
        f.write(f"**Total Duration:** {hours}h {minutes}m\n\n")
        f.write(f"## 📊 Statistics\n")
        f.write(f"- **Passed Specs:** {len(passed_specs)}\n")
        f.write(f"- **Failed Specs:** {len(failed_specs)}\n")
        f.write(f"- **Stuck/Idle Events:** {stuck_events}\n")
        f.write(f"- **Runner Restarts:** {restarts}\n\n")
        
        f.write("## ❌ Failed Specs\n")
        if not failed_specs:
            f.write("_No failures detected._\n")
        else:
            for spec in failed_specs:
                f.write(f"- `{spec}`\n")
        
        f.write("\n## ✅ Passed Specs (Sample)\n")
        for spec in passed_specs[-10:]: # Show last 10
            f.write(f"- `{spec}`\n")
            
    log(f"Report generated: {REPORT_FILE}")

def signal_handler(sig, frame):
    """Handles Ctrl+C gracefully."""
    global running
    log("⚠️ Interrupt received. Shutting down gracefully...")
    running = False

signal.signal(signal.SIGINT, signal_handler)

def stream_reader(pipe, prefix):
    """Reads stdout/stderr in a thread."""
    global last_output_time
    for line in iter(pipe.readline, b''):
        last_output_time = time.time()
        decoded_line = line.decode('utf-8', errors='replace').strip()
        if decoded_line:
            # Detect Pass/Fail (Adjust regex for your specific runner output)
            if "PASS" in decoded_line:
                # spec_name = ... extract if possible
                passed_specs.append(decoded_line)
            if "FAIL" in decoded_line:
                 # spec_name = ... extract if possible
                failed_specs.append(decoded_line)
                
            log(f"[{prefix}] {decoded_line}")
    pipe.close()

def run_tests():
    global process, last_output_time, restarts, log_cleaner
    
    # Start LogCleaner
    log_cleaner = LogCleaner(LOG_FILE)
    log_cleaner.start()
    
    while running:
        # Check total duration
        # rotate_log() -> Removed, handled by thread
        
        if time.time() - start_time > MAX_DURATION_SECONDS:
            log("🛑 6-Hour Time Limit Reached. Stopping.")
            break

        log(f"🚀 Starting Test Runner: {' '.join(TEST_COMMAND)}")
        last_output_time = time.time()
        
        try:
            # Start process
            if sys.platform == "win32":
                process = subprocess.Popen(TEST_COMMAND, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True)
            else:
                process = subprocess.Popen(TEST_COMMAND, stdout=subprocess.PIPE, stderr=subprocess.PIPE)

            # Start monitoring threads
            t_out = threading.Thread(target=stream_reader, args=(process.stdout, "OUT"))
            t_err = threading.Thread(target=stream_reader, args=(process.stderr, "ERR"))
            t_out.daemon = True
            t_err.daemon = True
            t_out.start()
            t_err.start()

            # Monitor loop
            while process.poll() is None and running:
                # Check idle timeout
                if time.time() - last_output_time > IDLE_TIMEOUT_SECONDS:
                    log(f"💤 Idle detected (> {IDLE_TIMEOUT_SECONDS/60} mins). Restarting runner...")
                    global stuck_events
                    stuck_events += 1
                    process.kill()
                    break # Break inner loop to restart
                
                # Check total duration in loop
                if time.time() - start_time > MAX_DURATION_SECONDS:
                    log("🛑 6-Hour Limit during execution. Killing process.")
                    process.terminate()
                    break

                time.sleep(1) # Prevent high CPU usage

            # Process finished
            exit_code = process.poll()
            if exit_code is not None:
                log(f"🏁 Runner exited with code {exit_code}")
                if exit_code != 0:
                    log(f"⚠️ Test failure or crash. Waiting {RETRY_DELAY_SECONDS}s before retry...")
                    restarts += 1
                    if restarts > MAX_RETRIES:
                        log("❌ Max retries reached. Aborting.")
                        break
                    time.sleep(RETRY_DELAY_SECONDS)
                else:
                    log("✅ Tests completed successfully.")
                    # Optional: Break if you only want ONE successful pass.
                    # For a "Watcher", maybe we want to run again? 
                    # The requirement says "orchestrate test lifecycle". Usually implies looped testing or 1 prolonged run.
                    # Assuming we want to KEEP running for 6 hours (e.g. repeated stress test?).
                    # If "Watcher" implies "Watch file changes", we rely on `npm test --watch`. 
                    # But here we are invoking `npm test`. 
                    # If it finishes, we'll wait and run again? Or exit?
                    # "After 6 hours... shut down". Implies it runs continuously.
                    # I will add a delay and re-run to fill the 6 hours.
                    log("🔄 Cycle complete. Restarting in 60s...")
                    time.sleep(60)

        except Exception as e:
            log(f"❌ Exception in runner: {e}")
            time.sleep(30)

    if log_cleaner:
        log_cleaner.stop()
    generate_report()
    log("👋 Antigravity Watcher shutdown complete.")

if __name__ == "__main__":
    log("=== Antigravity Watcher Started ===")
    run_tests()

File: innomcp-node/antigravity_watcher.js

const { spawn } = require('child_process');
const fs = require('fs');
const path = require('path');

// --- CONFIGURATION ---
const TEST_COMMAND_ARGS = ['test']; // "npm test" -> "npm" + ["test"]
const MAX_DURATION_MS = 6 * 3600 * 1000; // 6 Hours
const IDLE_TIMEOUT_MS = 10 * 60 * 1000; // 10 Minutes
const RETRY_DELAY_MS = 10 * 1000;       // 10 Seconds
const MAX_RETRIES = 3;                  // Max restarts

const LOG_FILE = 'antigravity_session.log';
const REPORT_FILE = 'summary_report.md';

// --- STATE ---
const startTime = Date.now();
let running = true;
let lastOutputTime = Date.now();
let passedSpecs = [];
let failedSpecs = [];
let stuckEvents = 0;
let restarts = 0;
let processRef = null;

function log(msg) {
    const timestamp = new Date().toISOString().replace('T', ' ').slice(0, 19);
    const line = `[${timestamp}] ${msg}`;
    console.log(line);
    fs.appendFileSync(LOG_FILE, line + '\n');
}

function rotateLog() {
    if (!fs.existsSync(LOG_FILE)) return;
    
    try {
        const data = fs.readFileSync(LOG_FILE, 'utf8');
        const lines = data.split('\n');
        
        if (lines.length > 10000) {
            console.log(`🧹 Log file exceeded 10,000 lines (${lines.length}). Truncating top 50%...`);
            // Keep last 5000 lines
            const newLines = lines.slice(lines.length - 5000);
            fs.writeFileSync(LOG_FILE, newLines.join('\n'));
            log(`✅ Log truncation complete. New size: ${newLines.length} lines.`);
        }
    } catch (err) {
        console.error(`⚠️ Log rotation failed: ${err.message}`);
    }
}

function generateReport() {
    const durationMs = Date.now() - startTime;
    const hours = Math.floor(durationMs / 3600000);
    const minutes = Math.floor((durationMs % 3600000) / 60000);

    let content = `# 🛡️ Antigravity Watcher Report (Node.js Edition)\n\n`;
    content += `**Date:** ${new Date().toLocaleString()}\n`;
    content += `**Total Duration:** ${hours}h ${minutes}m\n\n`;
    content += `## 📊 Statistics\n`;
    content += `- **Passed Specs:** ${passedSpecs.length}\n`;
    content += `- **Failed Specs:** ${failedSpecs.length}\n`;
    content += `- **Stuck/Idle Events:** ${stuckEvents}\n`;
    content += `- **Runner Restarts:** ${restarts}\n\n`;

    content += `## ❌ Failed Specs\n`;
    if (failedSpecs.length === 0) content += `_No failures detected._\n`;
    else failedSpecs.forEach(spec => content += `- \`${spec}\`\n`);

    content += `\n## ✅ Passed Specs (Sample)\n`;
    passedSpecs.slice(-10).forEach(spec => content += `- \`${spec}\`\n`);

    fs.writeFileSync(REPORT_FILE, content);
    log(`Report generated: ${REPORT_FILE}`);
}

function startRunner() {
    if (!running) return;
    
    rotateLog(); // Check log size before starting

    if (Date.now() - startTime > MAX_DURATION_MS) {
        log("🛑 6-Hour Time Limit Reached. Stopping.");
        running = false;
        generateReport();
        process.exit(0);
        return;
    }

    log(`🚀 Starting Test Runner: npm ${TEST_COMMAND_ARGS.join(' ')}`);
    lastOutputTime = Date.now();

    const isWin = process.platform === 'win32';
    // On Windows, use npm.cmd
    const cmd = isWin ? 'npm.cmd' : 'npm';
    
    processRef = spawn(cmd, TEST_COMMAND_ARGS, { shell: true });

    // Stream monitoring
    const onData = (data, source) => {
        lastOutputTime = Date.now();
        const lines = data.toString().split('\n');
        lines.forEach(line => {
            const trimmed = line.trim();
            if (!trimmed) return;
            
            // Detect Status
            if (trimmed.includes('PASS')) passedSpecs.push(trimmed);
            if (trimmed.includes('FAIL')) failedSpecs.push(trimmed);
            
            log(`[${source}] ${trimmed}`);
        });
    };

    processRef.stdout.on('data', d => onData(d, 'OUT'));
    processRef.stderr.on('data', d => onData(d, 'ERR'));

    processRef.on('close', (code) => {
        log(`🏁 Runner exited with code ${code}`);
        processRef = null;

        if (!running) return;

        if (code !== 0) {
            log(`⚠️ Failure detected. Waiting ${RETRY_DELAY_MS/1000}s...`);
            restarts++;
            if (restarts > MAX_RETRIES) {
                log("❌ Max retries reached. ABORTING.");
                running = false;
                generateReport();
                process.exit(1);
            } else {
                setTimeout(startRunner, RETRY_DELAY_MS);
            }
        } else {
            log("✅ Tests completed successfully.");
            log("🔄 Cycle complete. Restarting in 60s...");
            setTimeout(startRunner, 60000);
        }
    });

    processRef.on('error', (err) => {
        log(`❌ Spawn Error: ${err.message}`);
        setTimeout(startRunner, 30000);
    });
}

// Watchdog Loop
setInterval(() => {
    if (!running) return;
    
    // Check Global Timeout
    if (Date.now() - startTime > MAX_DURATION_MS) {
        log("🛑 Global Timeout. Killing process.");
        if (processRef) processRef.kill();
        running = false;
        generateReport();
        process.exit(0);
    }

    // Check Idle
    if (processRef && (Date.now() - lastOutputTime > IDLE_TIMEOUT_MS)) {
        log(`💤 Idle detected (> ${IDLE_TIMEOUT_MS/60000} mins). Restarting runner...`);
        stuckEvents++;
        processRef.kill(); 
        // 'close' handler will trigger restart
    }
}, 5000); // Check every 5s

// Cleanup
const cleanup = () => {
    if (running) {
        log("⚠️ Interrupt received. Shutting down...");
        running = false;
        if (processRef) processRef.kill();
        generateReport();
        process.exit(0);
    }
};
process.on('SIGINT', cleanup);
process.on('SIGTERM', cleanup);

log("=== Antigravity Watcher (Node.js) Started ===");
startRunner();
